### ISSUE TITLE
DL2

### ISSUE THREAD

**Issue by Dadabhainaoroji** [Dadabhainaoroji/Php#25]
1.     Design and implement a neural based network for generating word embedding for words in a document corpus 

!pip install -q gensim

from gensim.models import Word2Vec

 

corpus = [

    "The king and queen ruled the kingdom.",

    "A man and a woman are walking in the rain.",

    "Teachers share knowledge with their students."

]

 

tokenized = [s.lower().split() for s in corpus]

model = Word2Vec(tokenized, vector_size=50, window=3, min_count=1, sg=1, epochs=100)

 

for w in model.wv.index_to_key:

    print(f"{w}: {model.wv[w][:1]}")

 

 

2.     Write a program to demonstrate the working of a deep neural network for classification task. 

%pip install -q tensorflow matplotlib seaborn

import numpy as np

import pandas as pd

import matplotlib.pyplot as plt

import seaborn as sns

from tensorflow import keras

from sklearn.metrics import confusion_matrix, precision_recall_curve, auc, f1_score

from sklearn.preprocessing import label_binarize

 

 

(x_train, y_train), (x_test, y_test) = keras.datasets.mnist.load_data()

print("Train X:", x_train.shape)

print("Train Y:", y_train.shape)

print("Test X:", x_test.shape)

print("Test Y:", y_test.shape)

 

sampleNo = 1002

plt.imshow(x_train[sampleNo], cmap='gray')

print("Ground Truth Value:", y_train[sampleNo])

 

x_train = (x_train.astype("float32") / 255.0).reshape((-1, 28 * 28))

x_test = (x_test.astype("float32") / 255.0).reshape((-1, 28 * 28))

 

print(x_train.shape)

print(x_test.shape)

 

model = keras.Sequential([

    keras.layers.Input((28 * 28,)),

    keras.layers.Dense(512, activation="relu"),

    keras.layers.Dense(256, activation="relu"),

    keras.layers.Dense(128, activation="relu"),

    keras.layers.Dense(10, activation="softmax")

])

model.compile(optimizer="adam", loss="sparse_categorical_crossentropy", metrics=["accuracy"])

 

history = model.fit(x_train, y_train, batch_size=32, validation_split=0.1, epochs=15)

 

def plot_history(h):

    df = pd.DataFrame(h)

    df[['loss','val_loss']].plot()

    df[['accuracy','val_accuracy']].plot()

    plt.show()

 

plot_history(history.history)

 

y_pred_prob = model.predict(x_test[:5])

y_pred_classes = np.argmax(y_pred_prob, axis=1)

print("Predicted", y_pred_classes)

print("Ground Truth Value::", y_test[:5])

 

test_loss, test_acc = model.evaluate(x_test, y_test)

print(f"Test accuracy: {test_acc:.4f}")

 

y_predicted = model.predict(x_test)

y_predicted_classes = np.argmax(y_predicted, axis=1)

 

print("y_test shape:", y_test.shape)

print("y_predicted_classes shape:", y_predicted_classes.shape)

 

cm = confusion_matrix(y_test, y_predicted_classes)

plt.figure(figsize=(10,8))

sns.heatmap(cm, annot=True, fmt="d", cmap="Blues")

plt.xlabel("Predicted")

plt.ylabel("True")

plt.title("Confusion Matrix")

plt.show()

 

y_test_bin = label_binarize(y_test, classes=list(range(10)))

 

precision = {}

recall = {}

average_precision = {}

 

for i in range(10):

    precision[i], recall[i], _ = precision_recall_curve(y_test_bin[:, i], y_predicted[:, i])

    average_precision[i] = auc(recall[i], precision[i])

 

plt.figure(figsize=(10,8))

for i in range(10):

    plt.plot(recall[i], precision[i], lw=2, label=f'Class {i} (AP={average_precision[i]:0.2f})')

plt.xlabel("Recall")

plt.ylabel("Precision")

plt.title("Precision-Recall Curve per Class")

plt.legend(loc="lower left")

plt.show()

 

f1score = f1_score(y_test, y_predicted_classes, average='weighted')

print(f"F1 Score: {f1score:.4f}")

 

 

 

3.     Desing and implement a Convolutional Neural Network(CNN) for classification of image dataset 

%pip install -q tensorflow matplotlib

import tensorflow as tf

import keras

import os

import numpy as np

import pandas as pd

import matplotlib.pyplot as plt

import kagglehub

 

path = kagglehub.dataset_download("tongpython/cat-and-dog")

print("Path to dataset files:", path)

 

data_dir = '/kaggle/input/cat-and-dog/'

train_dir = os.path.join(data_dir, 'training_set/training_set')

test_dir = os.path.join(data_dir, 'test_set/test_set')

 

img_height = 128

img_width = 128

batch_size = 32

 

loader = dict(

    labels='inferred',

    label_mode='int',

    image_size=(img_height, img_width),

    interpolation='nearest',

    batch_size=batch_size

)

 

train_ds = tf.keras.utils.image_dataset_from_directory(train_dir, **loader)

test_ds = tf.keras.utils.image_dataset_from_directory(test_dir, **loader)

 

print(train_ds.class_names)

print(test_ds.class_names)

 

model = keras.Sequential([

    keras.Input(shape=(img_height, img_width, 3)),

    keras.layers.Conv2D(32, 3, padding='same', use_bias=False),

    keras.layers.BatchNormalization(),

    keras.layers.Activation('relu'),

    keras.layers.MaxPooling2D(2),

    keras.layers.Conv2D(64, 3, padding='same', use_bias=False),

    keras.layers.BatchNormalization(),

    keras.layers.Activation('relu'),

    keras.layers.MaxPooling2D(2),

    keras.layers.Conv2D(128, 3, padding='same', use_bias=False),

    keras.layers.BatchNormalization(),

    keras.layers.Activation('relu'),

    keras.layers.MaxPooling2D(2),

    keras.layers.Dropout(0.3),

    keras.layers.Conv2D(256, 3, padding='same', use_bias=False),

    keras.layers.BatchNormalization(),

    keras.layers.Activation('relu'),

    keras.layers.MaxPooling2D(2),

    keras.layers.Dropout(0.3),

    keras.layers.Flatten(),

    keras.layers.Dense(256, use_bias=False),

    keras.layers.BatchNormalization(),

    keras.layers.Activation('relu'),

    keras.layers.Dropout(0.5),

    keras.layers.Dense(1, activation='sigmoid')

])

 

model.summary()

model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

 

history = model.fit(train_ds, validation_data=test_ds, epochs=15)

model.save("CatDog_15Epoch.keras")

 

history_df = pd.DataFrame(history.history)

history_df[['loss','val_loss']].plot()

history_df[['accuracy','val_accuracy']].plot()

plt.show()

 

 

4.     Build and demonstrate an autoencoder network using neural layers for data compression on image dataset. 

%pip install -q tensorflow

import tensorflow as tf

from tensorflow.keras import layers, models

import matplotlib.pyplot as plt

import numpy as np

 

(X_train, _), (X_test, _) = tf.keras.datasets.mnist.load_data()

X_train = (X_train.astype('float32') / 255.0)[..., np.newaxis]

X_test = (X_test.astype('float32') / 255.0)[..., np.newaxis]

 

def build_autoencoder_model():

    encoder_input = layers.Input(shape=(28, 28, 1))

    x = layers.Conv2D(16, 3, padding='same', strides=2)(encoder_input)

    x = layers.BatchNormalization()(x)

    x = layers.Activation('relu')(x)

    x = layers.Conv2D(2, 3, activation='relu', padding='same', strides=2)(x)

    x = layers.Conv2DTranspose(16, 3, padding='same', strides=2)(x)

    x = layers.BatchNormalization()(x)

    x = layers.Activation('relu')(x)

    decoder_output = layers.Conv2DTranspose(1, 3, activation='sigmoid', padding='same', strides=2)(x)

    return models.Model(encoder_input, decoder_output)

 

model = build_autoencoder_model()

model.summary()

model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

 

model.fit(X_train, X_train, epochs=15, batch_size=128, shuffle=True, validation_data=(X_test, X_test))

 

def display_reconstruction(original, reconstructed):

    plt.figure(figsize=(10, 4))

    count = original.shape[0]

    for i in range(count):

        plt.subplot(2, count, i + 1)

        plt.imshow(original[i].squeeze(), cmap='gray')

        plt.axis('off')

        plt.subplot(2, count, count + i + 1)

        plt.imshow(reconstructed[i].squeeze(), cmap='gray')

        plt.axis('off')

    plt.show()

 

test_samples = X_test[:15]

reconstructed = model.predict(test_samples)

display_reconstruction(test_samples, reconstructed)

 

 

 

5.     Desing and implement a deep learning network for classification of textual documents. 

%pip install -q tensorflow

import tensorflow as tf

from tensorflow import keras

import numpy as np

import matplotlib.pyplot as plt

 

vocab_size = 10000

maxlen = 200

 

(x_train, y_train), (x_test, y_test)= keras.datasets.imdb.load_data(num_words=vocab_size)

 

x_train = keras.preprocessing.sequence.pad_sequences(x_train, maxlen=maxlen)

x_test = keras.preprocessing.sequence.pad_sequences(x_test, maxlen=maxlen)

 

print(f" Training samples: {len(x_train)}")

print(f"Testing samples: {len(x_test)}")

 

model = keras.Sequential([

    keras.layers.Embedding(input_dim=vocab_size, output_dim=64, input_length=maxlen),

    keras.layers.LSTM(128),

    keras.layers.Dense(64, activation='relu'),

    keras.layers.Dense(1, activation='sigmoid')

])

 

model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

model.build(input_shape=(None, maxlen))

model.summary()

 

history = model.fit(x_train, y_train, epochs=20, batch_size=128, validation_split=0.2, verbose=1)

model.save("TextDoc_Classification_model_1.keras")

 

test_loss, test_acc = model.evaluate(x_test, y_test, verbose=1)

print(f"Test accuracy: {test_acc:.3f}")

 

plt.plot(history.history['accuracy'], label='Train Accuracy')

plt.plot(history.history['val_accuracy'], label='Validation Accuracy')

plt.title('Training & Validation Accuracy')

plt.xlabel('Epochs')

plt.ylabel('Accuracy')

plt.legend()

plt.show()

 

word_index = keras.datasets.imdb.get_word_index()

 

def encode_text(text):

    tokens = [word_index.get(word.lower(), 2) for word in text.split()]

    return keras.preprocessing.sequence.pad_sequences([tokens], maxlen=maxlen)

 

sample_reviews = [

    "This movie was fantastic! I loved the acting and the story.",

    "Terrible film. Waste of time and money.",

    "It was ok, not great but can be watched not even once.",

]

 

for review in sample_reviews:

    encoded = encode_text(review)

    prediction = model.predict(encoded)[0][0]

    sentiment = "Positive" if prediction > 0.5 else "Negative"

    print(f"Review: {review}\nPredicted Sentiment: {sentiment} (Score: {prediction:.2f})\n")

 

 

6.     Design and implement a deep learning network for forecasting time series data.

%pip install -q yfinance

import numpy as np

import pandas as pd

import yfinance as yf

import matplotlib.pyplot as plt

from sklearn.preprocessing import MinMaxScaler

from tensorflow.keras.models import Sequential

from tensorflow.keras.layers import LSTM, Dense

 

data = yf.download('AAPL', start='2023-01-01', end='2024-01-01')

prices = data['Close'].values.reshape(-1,1)

 

scaler = MinMaxScaler()

scaled = scaler.fit_transform(prices)

 

def create_data(s, steps=20):

    X = np.array([s[i:i+steps,0] for i in range(len(s)-steps)])

    y = np.array([s[i+steps,0] for i in range(len(s)-steps)])

    return X, y

 

steps = 20

X, y = create_data(scaled, steps)

X = X.reshape((-1, steps, 1))

 

model = Sequential([LSTM(32, input_shape=(steps,1)), Dense(1)])

model.compile(optimizer='adam', loss='mse')

 

model.fit(X, y, epochs=15, batch_size=8, verbose=0)

 

pred_scaled = model.predict(X)

pred = scaler.inverse_transform(pred_scaled)

 

plt.plot(prices[steps:], label='True')

plt.plot(pred, label='Predicted')

plt.legend()

plt.show()

 

 

7.     Write a program to enable pre-train models to classify a given image dataset 

%pip install -q tensorflow

import tensorflow as tf

from tensorflow.keras import layers, models

from tensorflow.keras.applications import MobileNetV2

from tensorflow.keras.preprocessing import image

import numpy as np

import os

import matplotlib.pyplot as plt

import seaborn as sns

import kagglehub

from sklearn.metrics import confusion_matrix

 

train_ds = tf.keras.preprocessing.image_dataset_from_directory(

    '/kaggle/input/cat-and-dog/training_set/training_set',

    labels='inferred', label_mode='int',

    image_size=(224,224), interpolation='nearest', batch_size=32

)

 

test_ds = tf.keras.preprocessing.image_dataset_from_directory(

    '/kaggle/input/cat-and-dog/test_set/test_set',

    labels='inferred', label_mode='int',

    image_size=(224,224), interpolation='nearest', batch_size=32

)

 

base = MobileNetV2(weights='imagenet', include_top=False, input_shape=(224,224,3))

base.trainable = False

 

model = models.Sequential([

    base,

    layers.GlobalAveragePooling2D(),

    layers.Dense(512, use_bias="False"),

    layers.BatchNormalization(),

    layers.Activation('relu'),

    layers.Dropout(0.2),

    layers.Dense(128, use_bias="False"),

    layers.BatchNormalization(),

    layers.Activation('relu'),

    layers.Dropout(0.1),

    layers.Dense(len(train_ds.class_names), activation="softmax")

])

 

model.compile(optimizer="adam", loss="sparse_categorical_crossentropy", metrics=["accuracy"])

history = model.fit(train_ds, validation_data=test_ds, epochs=15)

model.save("PreTrainedCatDog.h5")

 

loss, acc = model.evaluate(test_ds)

print(acc)

 

path = kagglehub.dataset_download("tongpython/cat-and-dog")

img_path = os.path.join(path, "test_set/test_set/dogs/dog.4870.jpg")

img = image.load_img(img_path, target_size=(224,224))

np_img = image.img_to_array(img) / 255.0

img_array = np.expand_dims(np_img, axis=0)

prediction = model.predict(img_array)

print(test_ds.class_names[np.argmax(prediction)])

 

test_true = np.concatenate([y for x,y in test_ds], axis=0)

test_preds = model.predict(test_ds)

pred_labels = np.argmax(test_preds, axis=1)

 

cm = confusion_matrix(test_true, pred_labels)

plt.figure(figsize=(10,8))

sns.heatmap(cm, annot=True, fmt='d', cmap='Blues', xticklabels=test_ds.class_names, yticklabels=test_ds.class_names)

plt.show()

 

8.     Simple Grid World Problem: Design a custom 2D grid world where the agent navigates from a start position to a goal, avoiding obstacles. Environment: Custom grid (easily implemented in Python)

# LAB-8

from collections import deque

 

g = [["X",".",".",".","."],

     [".","X","X","G","."],

     [".",".",".","X","."],

     ["S",".",".",".","."]]

 

start, goal = (3,0), (1,3)

R,C = len(g), len(g[0])

dirs = {"up":(-1,0),"down":(1,0),"left":(0,-1),"right":(0,1)}

 

q, parent = deque([start]), {start: None}

while q:

    r,c = q.popleft()

    if (r,c)==goal: break

    for a,(dr,dc) in dirs.items():

        nr,nc = r+dr,c+dc

        if 0<=nr<R and 0<=nc<C and g[nr][nc]!="X" and (nr,nc) not in parent:

            parent[(nr,nc)] = ((r,c),a)

            q.append((nr,nc))

 

if goal not in parent:

    print("No path exists!")

else:

    path, cur = [], goal

    while parent[cur]:

        cur,act = parent[cur]; path.append(act)

    path.reverse()

    print("Path:", path)

 

    pos = start

    for a in path:

        pos = (pos[0]+dirs[a][0], pos[1]+dirs[a][1])

        print(pos)

**Comment by Shetty-Aryan** [comment]
I used this fix.
[4e77e546d776.zip](https://github.com/user-attachments/files/23697692/4e77e546d776.zip)
